### ISSUE TITLE
method called several times but verify just once with specific parameters

### ISSUE THREAD

**Issue by gambr** [mat007/turtle#120]
Hi,
in my test a setValue(string name, double value) method is called several times and I would like to verify it is called exactly once with specific parameters.

```
MOCK_EXPECT(serviceMock->setValue).exactly(1).with([](const std::string& name, double value) {
    return name.find("INPUT_HW_CUEING_AOR_AUTO") != std::string::npos && value == 1.0;
});
```
Since setValue() is called also other times in the same test with different arguments, the test fails.

I tried in vain with:
`MOCK_EXPECT(serviceMock->setValue).at_least(1).with(mock::any, mock::any);
`
before or after the other MOCK_EXPECT line but it doesn't work.

Is it possible to handle this case? If so, how?

Regards
Gianni

**Comment by Flamefire** [comment]
Great you were able to find the solution to your issue yourself.   
Checking the reported errors should have given enough information, so just to verify this is the case:

> Since setValue() is called also other times in the same test with different arguments, the test fails.

You mean it was giving you errors mentioning unexpected calls? This is the expected behavior as described in the "Getting started" section: https://github.com/mat007/turtle/blob/1796f40e586bbc581b42b69e0a1c8194db73cc7c/doc/getting_started.qbk#L43-L48

> I tried in vain with: `MOCK_EXPECT(serviceMock->setValue).at_least(1).with(mock::any, mock::any); ` before or after the other MOCK_EXPECT line but it doesn't work.

You didn't specify what "doesn't work". Putting this before your other expectation will basically consume any call. Putting it after will trigger a failed expectation if the method is called only once with the params expected by the other expectation. Again this is working as described and IMO reasonably expectable reading the code: 
```
MOCK_EXPECT(serviceMock->setValue).exactly(1).with(...)
MOCK_EXPECT(serviceMock->setValue).at_least(1).with(mock::any, mock::any);
```

So in the end your specification is incomplete:
> I would like to verify it is called exactly once with specific parameters.

This doesn't include what should happen for calls of other parameters. Are those expected? Should they happen or not happen? (Possibly) how often?

> But, if there isn't any different way to wite the same test, don't you think it would be nice to have a single statement to discard all other calls with different parameters?

If I understand this correctly you want to verify that the function is called with those exact parameters once and any number of times (including never) with other parameters?   
This is how you'd do that:
```
MOCK_EXPECT(serviceMock->setDatabaseValue).exactly(1).with([](const std::string& name, double value) {
      return name == "CF_LP_XY" && value == 1.0;
});
MOCK_EXPECT(serviceMock->setDatabaseValue);
```

The `with(mock::any, mock::any, ...)` is basically: Accept any parameter, i.e. no constraint(s). So removing it doesn't change the semantics. And `at_least(0)` is the explicit version of an unconstrained number of calls, see the documentation: https://github.com/mat007/turtle/blob/master/doc/reference.qbk#L312

If your code passes with 

    MOCK_EXPECT(serviceMock->setDatabaseValue).at_least(0).with(mock::any, mock::any); 
    MOCK_EXPECT(serviceMock->setDatabaseValue).exactly(1).with([](const std::string& name, double value) {
        return name == "CF_LP_XY" && value == 1.0;
    });

then there is likely a bug in Turtle as the 2nd expectation should never be fulfilled because the first (equivalent to `MOCK_EXPECT(serviceMock->setDatabaseValue)`) always matches. However I wasn't able to reproduce this in a test case so I guess you have the order the other way round which is the previous code I've shown.

Hope that helps.

**Comment by gambr** [comment]
Thanks Alexander for taking care of my request and trying to understand it. In fact you are almost completely right. The "doesn't work" meant "unexpected call" and my solution was not really working since all the calls were eaten by the first expectation but, if I understood correctly, also what you suggested does not work. Please read below.

> If I understand this correctly you want to verify that the function is called with those exact parameters once and any number of times (including never) with other parameters?

Yes, you are right, that's exactly what I was looking for.

> This is how you'd do that:
> 
> ```
> MOCK_EXPECT(serviceMock->setDatabaseValue).exactly(1).with([](const std::string& name, double value) {
>       return name == "CF_LP_XY" && value == 1.0;
> });
> MOCK_EXPECT(serviceMock->setDatabaseValue);
> ```

Unfortunately in this case even if the lambda returns false, the overall test succeeded because of the MOCK_EXPECT(serviceMock->setDatabaseValue) at the end. Let's say there is a bug in my production code and "value" checked in the lambda is not 1.0. That expectation fails returning false from the lambda but the test is wrongly successful.

I verified that. I comment out all of my setDatabaseValue calls from my production code except the one with first parameter = "CF_LP_XY", so that the MOCK_EXPECT(serviceMock->setDatabaseValue) should not be really needed. Anyway if I leave the MOCK_EXPECT(serviceMock->setDatabaseValue) the test at the very end is successful, while if I remove that line the test correctly fail (because the only expectation is the one returning false from the lambda).

So, at the moment I'm not able to check what you correctly understood: "want to verify that the function is called with those exact parameters once and any number of times (including never) with other parameters".

Best
Gianni

**Comment by Flamefire** [comment]
> Unfortunately in this case even if the lambda returns false, the overall test succeeded because of the MOCK_EXPECT(serviceMock->setDatabaseValue) at the end.

That is not possible because the expectation `MOCK_EXPECT(serviceMock->setDatabaseValue).exactly(1)` is not met.

So it looks like you didn't verify your expectations. I.e. specify when all expectations should have been met.

Verification happens implicitly when the mock-object is destroyed or explicitly when you call `mock::verify` on it. Given you are using pointer syntax here I guess your mock doesn't get destroyed at the end of the test. Fix that or use `mock::verify()`.

**Comment by gambr** [comment]
I'm a little confused. With mock::verify() I get the failing test correctly, but debugging the code the destructor of the mock-object is correctly called when BOOST_FIXTURE_TEST_CASE ends. Is this the correct flow? The mock-object is a shared_ptr created at the first line of the test, not a raw pointer, so I don't call delete explicitly. Can this make any difference?

**Comment by Flamefire** [comment]
Using a shared_ptr is fine. I'd guess you still have a reference to it alive so it isn't actually destroyed.

**Comment by gambr** [comment]
I set a breakpoint in the destructor of the mock-oject ad it is destroyed indeed. I tried to introduce a mem leak creating the object to test in the heap and verified in this case the mock-object destructor is never called. So, in my original case the mock-object is deleted for sure.

**Comment by Flamefire** [comment]
Destroying the mock object should also destroy the `mock::detail::function<T>` instances created via `MOCK_METHOD` which in turn triggers the verification of all set expectations, see `include/turtle/detail/function_impl.hpp` (unless there is an exception in flight)

Set a breakpoint in `~function_impl` to further debug this.

Note that `mock::detail::function<T>` acts as a reference type. Copying the mock object will also keep the function with its expectations in the copy (a shared_ptr), so verification only happens when the last copy is destroyed. Did you maybe create a copy of your mock outliving the test?

**Comment by mat007** [comment]
Hey @gambr,

Any chance you could come up with a simple reproduction? Basically you delete as much code as you code until it still shows the problem and you can share it in its entirety?

**Comment by gambr** [comment]
Sorry for the delay. I didn't have time to investigate where was the point but today I found how to create a simple example to reproduce case.
It seems related to std::unique_ptr pointing to base class. In my example below the behaviour is different if I use
1) std::unique_ptr< MyClass > myclass = std::make_unique< MyClass >(collaboratorMock);
with respect to
2) std::unique_ptr< IMyClass > myclass = std::make_unique< MyClass >(collaboratorMock);

Case 1) works as expected (the test fails). On the contrary, case 2) does not work as expected, causing the test to succeed even though it should fail.

Here is the code.

```
class MyCollaborator {
public:
    MyCollaborator() {}

    virtual void foo(double value) {}
};

MOCK_BASE_CLASS(MyCollaboratorMock, MyCollaborator)
{
    MOCK_METHOD(foo, 1)
};

class IMyClass {
public:
    virtual void doFoo() = 0;
};

class MyClass : public IMyClass {
public:
    MyClass(std::shared_ptr< MyCollaborator > collaborator)
    : collaborator(collaborator) {}

    void doFoo() override { collaborator->foo(1.0); }

private:
    std::shared_ptr< MyCollaborator > collaborator;
};

BOOST_AUTO_TEST_CASE(MyTest) {
    std::shared_ptr< MyCollaboratorMock > collaboratorMock = std::make_shared< MyCollaboratorMock >();

    MOCK_EXPECT(collaboratorMock->foo).once().with([&](double value) {
        return value == 2.0;
    });
    MOCK_EXPECT(collaboratorMock->foo).with(mock::any);

    //std::unique_ptr< MyClass > myclass = std::make_unique< MyClass >(collaboratorMock); //this works as expected: the test fails
    std::unique_ptr< IMyClass > myclass = std::make_unique< MyClass >(collaboratorMock);  //the test succeeds, while it should fail
    myclass->doFoo();
}
```

**Comment by Flamefire** [comment]
Try adding a virtual dtor to your interface as is best practice for virtual base classes/interfaces to avoid exactly issues like this:

```
class IMyClass {
public:
    virtual ~IMyClass() = default;
    virtual void doFoo() = 0;
};
```

The cause is pretty simple: Without this, a `delete IMyClass_instance` will not free any resources from the derived class, including the mock object.
